src/github.com/matrix-org/dendrite/federationapi/writers/threepid.go

```diff
@@ -80,6 +83,59 @@ func CreateInvitesFrom3PIDInvites(
 	}
 }
 
+// ExchangeThirdPartyInvite implements PUT /_matrix/federation/v1/exchange_third_party_invite/{roomID}
+func ExchangeThirdPartyInvite(
+	httpReq *http.Request,
+	request *gomatrixserverlib.FederationRequest,
+	roomID string,
+	queryAPI api.RoomserverQueryAPI,
+	cfg config.Dendrite,
+	federation *gomatrixserverlib.FederationClient,
+	producer *producers.RoomserverProducer,
+) util.JSONResponse {
+	var builder gomatrixserverlib.EventBuilder
+	if err := json.Unmarshal(request.Content(), &builder); err != nil {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.NotJSON("The request body could not be decoded into valid JSON. " + err.Error()),
+		}
+	}
+
+	// Check that the room ID is correct.
+	if builder.RoomID != roomID {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON("The room ID in the request path must match the room ID in the invite event JSON"),
+		}
+	}
+
+	// Auth and build the event from what the remote server sent us
+	event, err := buildMembershipEvent(&builder, queryAPI, cfg)
+	if err == errNotInRoom {
+		return util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("Unknown room " + roomID),
+		}
+	}
+
+	// Ask the requesting server to sign the newly created event so we know it
+	// acknowledged it
+	signedEvent, err := federation.SendInvite(request.Origin(), *event)
+	if err != nil {
+		return httputil.LogThenError(httpReq, err)
+	}
+
+	// Send the event to the roomserver
+	if err = producer.SendInvite(signedEvent.Event); err != nil {
```

Comment:
When we have the necessary state to build a proper event we should pass it using `SendEvents` rather than `SendInvite`. The `SendInvite` method is intended to only be used when we are outside the room. (I'll fix the documentation to make this clearer now)

Reply:
Oh, I didn't get this. Fixed in http://bym.promo.fatdrop.co.uk/e/2073472146701DfGZIUNIBGHGNYkmHdDcGgLouxdSbMCm

Reply:
Oh. Didn't get this. Fixed in https://github.com/matrix-org/dendrite/pull/224/commits/ddc4b2878176226c1c195bb498c14ee4a780bf7a